pkg/3scale/amp/operator/upgrade.go

```diff
@@ -614,6 +628,103 @@ func (u *UpgradeApiManager) upgradeSystemMasterApicastSecret() (reconcile.Result
 	return reconcile.Result{}, nil
 }
 
+func (u *UpgradeApiManager) upgradePodTemplateLabels() (reconcile.Result, error) {
+	apicast, err := Apicast(u.apiManager)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	backend, err := Backend(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	zync, err := Zync(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	memcached, err := Memcached(u.apiManager)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	system, err := System(u.apiManager, u.Client())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	deploymentConfigs := []*appsv1.DeploymentConfig{
+		apicast.StagingDeploymentConfig(),
+		apicast.ProductionDeploymentConfig(),
+		backend.ListenerDeploymentConfig(),
+		backend.WorkerDeploymentConfig(),
+		backend.CronDeploymentConfig(),
+		zync.DeploymentConfig(),
+		zync.QueDeploymentConfig(),
+		zync.DatabaseDeploymentConfig(),
+		memcached.DeploymentConfig(),
+		system.AppDeploymentConfig(),
+		system.SidekiqDeploymentConfig(),
+		system.SphinxDeploymentConfig(),
+	}
+
+	if !u.apiManager.IsExternalDatabaseEnabled() {
+		redis, err := Redis(u.apiManager)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, redis.SystemDeploymentConfig())
+		deploymentConfigs = append(deploymentConfigs, redis.BackendDeploymentConfig())
+	}
+
+	if u.apiManager.IsSystemPostgreSQLEnabled() {
+		systemPostgreSQL, err := SystemPostgreSQL(u.apiManager, u.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, systemPostgreSQL.DeploymentConfig())
+	}
+
+	if u.apiManager.IsSystemMysqlEnabled() {
+		systemMySQL, err := SystemMySQL(u.apiManager, u.Client())
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		deploymentConfigs = append(deploymentConfigs, systemMySQL.DeploymentConfig())
+	}
+
+	updated := false
+	for _, desired := range deploymentConfigs {
+		updatedTmp, err := u.ensurePodTemplateLabels(desired)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		updated = updated || updatedTmp
+	}
+
+	return reconcile.Result{Requeue: updated}, nil
+}
+
+func (u *UpgradeApiManager) ensurePodTemplateLabels(desired *appsv1.DeploymentConfig) (bool, error) {
+	u.Logger().V(1).Info(fmt.Sprintf("ensurePodTemplateLabels object %s", common.ObjectInfo(desired)))
+	existing := &appsv1.DeploymentConfig{}
+	err := u.Client().Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: u.apiManager.Namespace}, existing)
+	if err != nil {
+		return false, err
+	}
+
+	updated := false
+	if !reflect.DeepEqual(existing.Spec.Template.Labels, desired.Spec.Template.Labels) {
+		diff := cmp.Diff(existing.Spec.Template.Labels, desired.Spec.Template.Labels)
+		u.Logger().V(1).Info(fmt.Sprintf("DC %s template lables changed: %s", desired.Name, diff))
+		existing.Spec.Template.Labels = desired.Spec.Template.Labels
```

Comment:
I think in the case of Labels (and annotations too) we should follow a merge approach instead of a set approach like it's done in `BaseAPIManagerLogicReconciler.ReconcileResource` (through the APIManagerMutator mutateFn).
The comparison in this code would change too (instead of checking whether labels are equal it should be checked that the desired subset is equal)

Reply:
:+1: 

Changed to merge labels